No Slotting Arithmetic

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#94

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [ ] I have remove slotting arithmetic operation from the source code (relying instead on the fact that blocks are linked to each other and points to their parent)
- [ ] I have added a few tests to make sure that ordering of slots happens as expect, in a sudden paranoia crisis (since we derive the `Ord` automatically, it _just happened_ that we defined the epoch index before the slot number in the record declaration. Swapping those two would have dramatic effects ^^")

# Comments

<!-- Additional comments or screenshots to attach if any -->

As a side-effect, we don't have to test for any of the boundary cases because there are none :tada: 

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ commented on test/integration/Test/Integration/APISanityChecks.hs:
```diff
@@ -0,0 +1,132 @@
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Test.Integration.APISanityChecks
```
Pretty much copy-pasted from before. I only remove one check saying that all `DELETE` should be `204` (that's actually not true since some of our DELETE are `404`) and also increased the number of randomly generated transactions from 100 to 1000... 
I'd actually like to see what kind of requests are generated and if this whole thing actually does check something :thinking:  ... but for now, I just kept it.

KtorZ commented on test/integration/Test/Integration/Framework/Orphan.hs:
```diff
@@ -0,0 +1,37 @@
+numForbiddenOperation :: Text
+numForbiddenOperation =
+    "Call forbidden operation: 'Num' for 'Coin' is only defined to ease\
+    \ literal notation when writing integration tests scenarios."
```
Perhaps a `newtype` here.. I don't know

parsonsmatt replied:
Why not just use a constructor? or a function? I'm not a fan of abusing `Num` for things like this, where *all* the methods are errors.

KtorZ replied:
Well, the focus is really on keeping the end-to-end scenarios simple (read as: for non-haskellers), so I am really trying to remove as much noise as possible.

parsonsmatt replied:
Confusion can come from many things -- excess polymorphism can certainly do it. `coins 0` is slightly noisier than `0` but is more explicit and monomorphic, which should provide for better error messages too.

KtorZ commented on test/integration/Test/Integration/Framework/Request.hs:
```diff
@@ -0,0 +1,61 @@
+    request
+        :: forall m ctx. (MonadIO m, MonadReader ctx m, HasHttpClient ctx)
+        => (WalletClient IO -> IO (Either ClientError originalResponse))
+        -> m (Either ClientError (Response originalResponse))
```
I wonder if we couldn't modify the client to actually do that for us directly. The `jsend` wrapper is pretty much useless (except for the metadata maybe .. and even... ).

parsonsmatt replied:
The metadata is mostly useful for pagination

KtorZ commented on test/integration/Test/Integration/Framework/Scenario.hs:
```diff
@@ -0,0 +1,60 @@
+        runAndPersist mvar = do
+            s <- takeMVar mvar
+            (_, s') <- runStateT io s
+            putMVar mvar s'
```
So, here's the trick. We save the context from a scenario to another. The only thing that actually gets modified in the context are the faucet wallets. So here, we're basically keeping track of a counter, and it gets incremented between each test, like a good global variable. In theory, the tests themselves only need this to be a `MonadReader`.

KtorZ commented on test/integration/Test/Integration/Framework/Scenario.hs:
```diff
@@ -0,0 +1,60 @@
+            putMVar mvar s'
+
+-- | Execute all scenarios in sequence providing the given context 'ctx'
+runWithContext :: forall ctx. ctx -> SpecM ctx () -> Spec
```
So, HSpec can't run anything else than `Spec = SpecM (ActionWith ()) ()`; So this makes the transformation, provided a context.

parsonsmatt replied:
This can also be defined as `runWithContext ctx = beforeAll (pure ctx)`

KtorZ commented on test/integration/Test/Integration/Scenario/Transactions.hs:
```diff
@@ -0,0 +1,172 @@
+        verify response
+            [ expectTxInHistoryOf wallet
+            , expectTxStatusEventually [Creating, Applying, InNewestBlocks, Persisted]
+            ]
```
And finally, some test scenario ! Notice they're all divided in 3 "paragraphs"

- Setup
- Action
- Expectations

The DSL makes is possible to almost read it like a sentence and combinators are re-usable between types to make things quite smooth.

parsonsmatt commented on .stylish-haskell.yaml:
```diff
@@ -222,7 +222,9 @@ language_extensions:
   - DefaultSignatures
   - DeriveDataTypeable
   - DeriveGeneric
+  - DerivingStrategies
```
:joy:

parsonsmatt commented on cardano-wallet.cabal:
```diff
@@ -345,87 +345,6 @@ executable cardano-generate-swagger-file
-      Util
-      WalletSpecs
-
-
```
:joy:

parsonsmatt commented on test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -0,0 +1,859 @@
+-- | Just a slightly-specialized alias for 'it' to help lil'GHC. Also, makes
+-- sure the wallet has been cleared out completely before running the scenario.
+scenario
+    :: (HasCallStack, Example (Scenario Context IO ()))
```
This constraint is dodgy. It is fully saturated, which means we expect some orphan instance to come by and satisfy it.

KtorZ replied:
Hmmm... right my bad, there's no need for a constraint actually. I just "copy/pasted" the signature of `it` without much thinking :man_shrugging:

parsonsmatt commented on test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -0,0 +1,859 @@
+
+import           Control.Concurrent (threadDelay)
+import           Control.Concurrent.Async (Async, async, race, wait)
+import           Data.Generics.Product.Fields (HasField, field)
```
`generic-lens` <3 <3 <3

parsonsmatt commented on test/integration/Test/Integration/Framework/Scenario.hs:
```diff
@@ -0,0 +1,60 @@
+
+-- | A Wrapper around 'StateT' around which we define a few instances. The most
+-- interesting one is 'Example'
+newtype Scenario s m a = Scenario (StateT s m a)
```
Beware -- `StateT` is single-threaded state, and this is usually not what folks want in th epresence of concurrency and exceptions.

KtorZ replied:
Indeed, though here all the tests are running sequentially in a single-thread. We _could_ possibly run them in parallel I believe, and it shouldn't matter much because we don't do any concurrent business inside the `StateT`.
It'd would require to adjust where the `MVar` is manipulated however; but that's clearly not the intent anyway or at least, not in the first iteration :)

parsonsmatt commented on test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -0,0 +1,855 @@
+    :: (HasCallStack, Typeable s)
+    => Proxy s
+    -> a
+notInitialized proxy =
```
I'm feeling *really* iffy on this. It's used in 12 other places, and basically means that we're throwing away the "If it type checks, it works!" feature for Haskell for *anything* that calls one of the functions. I don't like that.

It seems that we're using this for almost all of the fields in all of the use sites. We only provide a few default values. It would be more direct to have the caller pass in the entire data structure -- right now, uses are using lensy setters (eg `walletId wallet :: HasType WalletId s => s -> s`). A more direct construction would use:

```haskell
getAccountBalance :: AccountBalance -> ...

getAccountBalance AccountBalance { _walletId = walId wallet } 
```

Then, with strict fields, the user gets an error if they forget to set something.

If we want to allow defaults, then we can use functions of the form `(defParam -> defParam2 -> Thing)`, like:

```haskell
getAccountBalance (\defWalletId defOtherThing -> AccountBalance myThing defWalletId defOtherThing) ...
```

The least boilerplate-y thing might just be to take the required things as parameters directly.

I'm happy to help with hashing out an alternative to this.

KtorZ replied:
I am rather unsatisfied as well about this one and I knew you'd point that out :upside_down_face:  ... It's really about the end goal in the end: having the end scenario written in the most simple and readable way. Ideally, I'd like to add something at the type-level in the data-structure to mark which fields are required in such way that the whole object becomes a monoid. Fields can be either defaulted to an empty value, or we use this "Required" type-wrapper to initialize the non-defaultable fields. 

```hs
data RequestParameters = RequestParameters
    { myField :: Required String
    , otherField :: [Whatever]
    }
```

Meh, I'll try to get something better :+1:

parsonsmatt replied:
There's a concept called Monoid Actions, with a class:

```haskell
class Action m a where
  act :: m -> a -> a
```

where `m` is a Monoid or Semigroup and the laws are basically:

```
act a . act b = act (a <> b)
```

This gives a nice abstraction on "partial configurations acting on a default" where a partial configuration has an "override" `<>` operation.

However, the abstraction ends up being kind-of a lot of boilerplate for not much gain over just being specific about what you mean.

parsonsmatt commented on test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -0,0 +1,855 @@
+
+fromRight :: Either e a -> a
+fromRight =
+    either (error "fromRight: can't take right side of 'Left' value") id
```
I would really prefer to see this as a `MonadFail` rather than a pure exception. There's only one use, and it's to set a backup phrase -- we *really* shouldn't be throwing away Haskell's type safety just because we're writing tests.

parsonsmatt commented on test/integration/Test/Integration/Framework/Request.hs:
```diff
@@ -0,0 +1,60 @@
+import           Test.Integration.Framework.Orphan ()
+
+
+type HasHttpClient ctx = HasField "_client" ctx ctx (WalletClient IO) (WalletClient IO)
```
note that `HasType (WalletClient IO) ctx` also works here and is more general :)

parsonsmatt replied:
I'm of the opinion that a class like this could also be a class alias:

```haskell
class (HasType (WalletClient IO) ctx) => HasHttpClient ctx where
    httpClient :: Lens' ctx (WalletClient IO)
    httpClient = typed @(WalletClient IO)

instance (HasType (WalletClient IO) ctx) => HasHttpClient ctx
```

This allows someone to import `HasHttpClient(..)` and get the type *and* lens imported at the same time.

parsonsmatt commented on test/integration/Test/Integration/Framework/Request.hs:
```diff
@@ -0,0 +1,60 @@
+instance Request (WalletResponse a) where
+    type Response (WalletResponse a) = a
+    request action = do
+        httpClient <- asks (^. #_client)
```
note that `view` works in a `MonadReader`, so this is also:


```suggestion
        httpClient <- view #_client
```

parsonsmatt commented on test/integration/Test/Integration/Framework/Scenario.hs:
```diff
@@ -0,0 +1,60 @@
+instance Example (Scenario s IO ()) where
+    type Arg (Scenario s IO ()) = MVar s
+    evaluateExample (Scenario io) _ action _ =
+        action runAndPersist >> return (Result "" Success)
```
It seems *really* wrong to me to unconditionally return `Result "" Success`. Don't we care about the result of `action`?

parsonsmatt replied:
Interesting.

Turns out `hspec` and the underlying libraries just assume an `IO` action completing at all is a success, and all test failures are done by exceptions. So if `action runAndPersist` throws, then `return` never executes!

parsonsmatt commented on test/integration/Test/Integration/Scenario/Transactions.hs:
```diff
@@ -0,0 +1,172 @@
+
+        response <- postTransaction
+            $ source wallet
+            . destination (addr, 14)
```
Now that I know these functions are all unsafe, I have an urge to check the definition of every single one and verify that you didn't forget to set the required parameters. It would be *much* easier and simpler if this were simply:

```haskell
response <- postTransaction
    (source wallet)
    (destination (addr, 14))
```

(or whatever the right lensy thing is)

parsonsmatt commented on test/integration/Test/Integration/Scenario/Wallets.hs:
```diff
@@ -0,0 +1,90 @@
+module Test.Integration.Scenario.Wallets
+  ( spec
+  ) where
```
2 space indent

KtorZ commented on test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -0,0 +1,855 @@
+    :: String
+    -> Scenario Context IO ()
+    -> Scenarios Context
+scenario title spec = it title (successfulRequest Client.resetWalletState >> spec)
```
True :) and to be honest, I thought about it. Though, this is a `before` we always want and can easily forget, hence it being shoved and hidden up here :stuck_out_tongue:

KtorZ commented on test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -0,0 +1,855 @@
+    -> Either ClientError a
+    -> m ()
+expectEqual =
+    expectFieldEqual ($)
```
Haha, true that :+1:

KtorZ commented on test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -0,0 +1,656 @@
+--
+-- DEFAULT VALUES
+--
+
```
All those default are quite "verbose" but they make the scenarios nicer to read in the end :+1:

KtorZ commented on test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -0,0 +1,656 @@
+--
+-- HELPERS
+--
+
```
@parsonsmatt 
I'd appreciate your inputs on the ones below :upside_down_face: Some are maybe a bit ad-hoc.

parsonsmatt commented on test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -0,0 +1,656 @@
+
+lens :: (s -> a) -> ((s, b) -> t) -> Lens s t a b
+lens _get _set = \f x -> curry _set x <$> f (_get x)
+{-# INLINE[0] lens #-}
```
why aren't we just using `Control.Lens`? the library already depends on it iirc

KtorZ replied:
Yeah, that was to have the definition in front of me while I was grooming haha. This is the kind of polish i told you about :upside_down_face:

parsonsmatt commented on test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -0,0 +1,656 @@
+    _get :: HasType (V1 Coin) s => s -> Word64
+    _get = unsafeGetCoin . unV1 . view typed
+    _set :: HasType (V1 Coin) s => (s, Word64) -> s
+    _set (s, v) = set typed (V1 $ mkCoin v) s
```
lgtm

parsonsmatt commented on test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -0,0 +1,656 @@
+    :: HasType Text s
+    => Lens' s Text
+walletName =
+    typed
```
For definitions as short as these, I'd probably condense to two lines:

```haskell
walletName :: HasType Text s => Lens' s Text
```

Note, however, that `walletName` will only work on data structures that only contain a *single* value of type `Text`, and will work on values like:

```haskell
data User = User { userName :: Text } deriving Generic
```

which is not what you want. So a `newtype WalletName = WalletName Text` might be warranted here?

KtorZ replied:
Yup, thought of it and we'll probably end up doing that. I didn't want to clutter the PR with type changes in our source code but well, that's next for sure (including removing all those V1)

KtorZ committed:
remove slotting arithmetic from source code

The rational is that, doing slotting arithmetic forces us to keep track of the 'slotsPerEpoch'
which is in practice, rather tricky. We can _probably_ get away by leveraging the _chained_
aspect of the blockchain (that each block points to its previous one) and rewind the chain
when needed.

KtorZ committed:
add extra tests to make sure the 'Ord' instance on slot does what we think it does

piotr-iohk commented on test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:
```diff
@@ -339,15 +291,10 @@ instance Arbitrary BlockHeader where
-        <$> choose (0 :: Int, 4 * (fromIntegral slotsPerEpoch))
+    arbitrary = do
+        ep <- choose (0, 10)
+        sl <- choose (0, 100)
```
why not `sl <- choose (0, 21600)`?
Also is it possible to have slot 0? or epoch 0?
It is because it's just an Id I suppose?

piotr-iohk replied:
hm, is it even needed? I mean the whole generator for `SlotId`? The [`Slotting ordering`](https://github.com/input-output-hk/cardano-wallet/pull/126/files#diff-00fd2cdc39324db62230aba60bf1287bR87) test suite ^ is using only some explicit values for SlotId...

KtorZ replied:
Not only, there's a first property that compare arbitrary slots against the initial `SlotId 0 0` (making sure they're all greater than it).

Why not `21600` ? Well, to really show that it doesn't matter here and is completely irrelevant for the testing of this module. The only thing we really test is that we compare slot correctly. There's no other assumption about `SlotId` (and that's what we try to preserve in the code as well).

piotr-iohk reviewed (approved):
:+1:
